agents/holo_synth/synth_agent.py

```diff
@@ -0,0 +1,208 @@
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        agg_params = {"frame_length": 10 * 60}  # [sec]
+        self.agent.register_feed(
+            "synth_LO", record=True, agg_params=agg_params, buffer_time=0
+        )
+        """
+        if mode == 'acq':
+            self.auto_acq = True
+        else:
+            self.auto_acq = False
+        self.sampling_frequency = float(samp)
+
+        ### register the position feeds
+        """
+        if config_file == "None":
+            raise Exception("No config file specified for the FTS mirror config")
+        else:
+            config_file_path = os.path.join(os.environ["OCS_CONFIG_DIR"], config_file)
+            print(config_file_path)
+            with open(config_file_path) as stream:
+                self.holog_configs = yaml.safe_load(stream)
+                if self.holog_configs is None:
+                    raise Exception("No mirror configs in config file.")
+                self.log.info(f"Loaded mirror configs from file {config_file_path}")
+                self.N_MULT = self.holog_configs.pop("N_MULT", None)
+                self.ghz_to_mhz = self.holog_configs.pop("ghz_to_mhz", None)
+
+    def init_synth(self, session, params=None):
+        """init_synth(params=None)
+        Perform first time setup for communication with Synth.
+        """
+
+        if params is None:
+            params = {}
+
+        self.log.debug("Trying to acquire lock")
+        with self.lock.acquire_timeout(timeout=0, job="init") as acquired:
+            # Locking mechanism stops code from proceeding if no lock acquired
+            if not acquired:
+                self.log.warn(
+                    "Could not start init because {} is already running".format(
+                        self.lock.job
+                    )
+                )
+                return False, "Could not acquire lock."
+            # Run the function you want to run
+            self.log.debug("Lock Acquired Connecting to Stages")
+            self.lo_id = synth3.get_LOs()
+
+            synth3.set_RF_output(0, 1, self.lo_id) # LO ID, On=1, USB connection ID
+            synth3.set_RF_output(1, 1, self.lo_id) # LO ID, On=1, USB connection ID
+
+            # data = {"timestamp": time.time(), "block_name": "synth_LO", "data": {}}
+
+            # data["data"]["F1_status"] = 1
+            # data["data"]["F2_status"] = 1
+
+            # self.agent.publish_to_feed("synth_LO", data)
+            # session.data.update(data["data"])
+
+
+        # This part is for the record and to allow future calls to proceed,
+        # so does not require the lock
+        self.initialized = True
+
+        return True, "Synth Initialized."
+
+    def set_frequencies(self, session, params):
+        """
+        params: 
+            dict: {'freq0': float
+                   'freq1': float}
+        """
+
+        f1 = params.get("freq1", 0)
+        f_offset = params.get("offset", 0)
```

Comment:
This is a good place to use the ocs param decorator. You can validate inputs so that users on the Client side don't do things like input a negative frequency, or exceed a maximum. Details here: https://ocs.readthedocs.io/en/develop/developer/agents.html#validation-using-param

It would also be good to note that the units are in GHz here.

Reply:
done in synth_agent for both tasks